### DISCUSSION TITLE
Exclude code and dependencies from the final version of the game

### DISCUSSION THREAD

**Discussion by DomeAlessi** [nxrighthere/UnrealCLR#238] [Help] [answered]
Hello I am new to Unreal (sorry for the noob question).

I would like to use UnrealCLR to develop (and port my old tools written in C#) in Unreal, I was wondering if I use UnrealCLR only to make tools and maybe depend on third party .Net libs how much of it and these libraries will be in the final version of the game?
So, in general, to what extent I can use UnrealCLR to benefit of the C# features but avoid to have any of this in the final package?

Thank you!
Domenico

**Comment by nxrighthere** [comment, marked as answer]
You can ship any 3rd-party libraries with the final version of the game as long as it's compatible with the plugin. Certain 3rd-party dependencies don't support [unloadability](https://docs.microsoft.com/en-us/dotnet/standard/assembly/unloadability), so they might block hot reload.

**Reply by DomeAlessi** [reply]
Thanks for your answer by re-reading my question i realized it was badly written, sorry :(
Can I use UnrealCLR just to make my tools in C# and exclude it from the final release of the game so that is not part of it at all?
(that's what i wanted to ask :D)

**Reply by nxrighthere** [reply]
The optimal solution for this is [DefineConstants](https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/compiler-options/language#defineconstants) option. You can compile relevant code with dependencies for the editor and final game separately. Use `#if`, `#else`, `#elif`, and `#endif` in source code. For package references in `csproj` use:
```csproj
<ItemGroup Condition="$(DefineConstants.Contains('EDITOR'))">
  <PackageReference ... />
</ItemGroup>
```
Then just compile assemblies with `-p:DefineConstants=EDITOR`. You can use your IDE to configure this.

**Reply by nxrighthere** [reply]
Also, you can make a runtime check for yourself with a trick mix of DefineConstants with [`Engine.IsEditor`](https://github.com/nxrighthere/UnrealCLR/blob/master/API/Engine-IsEditor.md), so you can throw an exception when editor code accidentally ends up in a final release depending on defined constants.

**Reply by DomeAlessi** [reply]
This is exactly what I needed to know, very much!
